### PR TITLE
PLT-3193 Add channel notification preferences for push and email noti…

### PR DESCRIPTION
**Pull request by enahum**
…fications

Please make sure you've read the [pull request](http://docs.mattermost.com/developer/contribution-guide.html#preparing-a-pull-request) section of our [code contribution guidelines](http://docs.mattermost.com/developer/contribution-guide.html).

When filling in a section please remove the help text and the above text.

#### Summary
This PR adds server side functionality for a channel to override the global notification settings for email and push notifications

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-3193

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [x] Has enterprise changes (https://github.com/mattermost/enterprise/pull/105)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (notifications and performance)

### CODE REVIEW

File: store/sql_channel_store.go [hunk 1]
```diff
@@ -919,6 +924,71 @@ func (s SqlChannelStore) GetAllChannelMembersForUser(userId string, allowFromCac
...
+			if metrics != nil {
+				metrics.IncrementMemCacheMissCounter("All Channel Members Notify Props for Channel")
+			}
+		}
+
+		var data []allChannelMemberNotifyProps
+		_, err := s.GetReplica().Select(&data, `
+			SELECT ChannelMembers.UserId, ChannelMembers.NotifyProps
+			FROM Channels, ChannelMembers
+			WHERE Channels.Id = ChannelMembers.ChannelId AND ChannelMembers.ChannelId = :ChannelId AND Channels.DeleteAt = 0`, map[string]interface{}{"ChannelId": channelId})
```
Comment (crspeller) [inline store/sql_channel_store.go]: Why do we need to check to make sure the channel is not deleted here?

File: store/sql_channel_store.go [hunk 2]
```diff
@@ -919,6 +924,71 @@ func (s SqlChannelStore) GetAllChannelMembersForUser(userId string, allowFromCac
...
+		} else {
+
+			props := make(map[string]model.StringMap)
+			for i := range data {
+				props[data[i].UserId] = data[i].NotifyProps
+			}
+
+			result.Data = props
+
+			if allowFromCache {
```
Comment (crspeller) [inline store/sql_channel_store.go]: It is not nessiary to check for this here. (yes I know we do it in some places)

File: store/sql_channel_store.go (hunk 1 above)
Comment (enahum) [inline store/sql_channel_store.go]: it doesn't hurt, but I can remove that from the query

File: store/sql_channel_store.go (hunk 2 above)
Comment (enahum) [inline store/sql_channel_store.go]: yes it should always check for that either way

**Comment by enahum** [comment]
@crspeller done with the changes, can I ask you to please review this PR https://github.com/mattermost/enterprise/pull/105 too.

Thanks

**Comment by crspeller** [comment]
@enahum Sorry I didn't mean to remove the cache check entirely I meant you don't need to check before adding it to the cache, only retrieving from.

**Comment by enahum** [comment]
@crspeller it should be good now

File: app/notification.go [hunk 3]
```diff
@@ -442,14 +455,22 @@ func GetMessageForNotification(post *model.Post, translateFunc i18n.TranslateFun
...
-func sendPushNotification(post *model.Post, user *model.User, channel *model.Channel, senderName string, wasMentioned bool) *model.AppError {
+func sendPushNotification(post *model.Post, user *model.User, channel *model.Channel, senderName string, channelNotifyProps model.StringMap, wasMentioned bool) *model.AppError {
 	sessions, err := getMobileAppSessions(user.Id)
 	if err != nil {
 		return err
 	}
 
 	var channelName string
 
+	if channelNotify, ok := channelNotifyProps["push"]; ok {
```
Comment (coreyhulen) [inline app/notification.go]: Can we make `"push"` a constant.

Comment (enahum) [inline app/notification.go]: yes I can, the same for email I guess

Comment (enahum) [inline app/notification.go]: @coreyhulen any particular reason why just "push" and not the other 3?

**Comment by coreyhulen** [comment]
Should be all 3.

**Comment by enahum** [comment]
@coreyhulen move them all to be constants

File: model/channel_member.go [hunk 4]
```diff
@@ -100,6 +100,20 @@ func (o *ChannelMember) IsValid() *AppError {
 			nil, "mark_unread_level="+markUnreadLevel)
 	}
 
+	if pushLevel, ok := o.NotifyProps["push"]; ok {
```
Comment (jwilander) [inline model/channel_member.go]: I think you missed using a constant here

File: model/channel_member.go [hunk 5]
```diff
@@ -100,6 +100,20 @@ func (o *ChannelMember) IsValid() *AppError {
...
+		if len(pushLevel) > 20 || !IsChannelNotifyLevelValid(pushLevel) {
+			return NewLocAppError("ChannelMember.IsValid", "model.channel_member.is_valid.push_level.app_error",
+				nil, "push_notification_level="+pushLevel)
+		}
+	}
+
+	if sendEmail, ok := o.NotifyProps["email"]; ok {
```
Comment (jwilander) [inline model/channel_member.go]: Here as well

**Comment by enahum** [comment]
@jwilander ready, also please don't forget to review the EE PR

**Pull request merged**